### PR TITLE
MaxHumeShield

### PR DESCRIPTION
**Pull request by louis1706**

### CODE REVIEW

File: Exiled.API/Features/Player.cs [hunk 1]
```diff
@@ -191,6 +192,7 @@ private set
                 CameraTransform = value.PlayerCameraReference;
 
                 value.playerStats._dictionarizedTypes[typeof(HealthStat)] = value.playerStats.StatModules[0] = healthStat = new CustomHealthStat { Hub = value };
+                value.playerStats._dictionarizedTypes[typeof(HumeShieldStat)] = value.playerStats.StatModules[0] = humeshieldStat = new CustomHumeShieldStat { Hub = value };
```
Comment (Thundermaker300) [inline Exiled.API/Features/Player.cs:195]:
```suggestion
                value.playerStats._dictionarizedTypes[typeof(HumeShieldStat)] = value.playerStats.StatModules[4] = humeshieldStat = new CustomHumeShieldStat { Hub = value };
```
Stat 4 is the hume shield stat. Stat 0 is the health stat.

File: Exiled.API/Features/Player.cs [hunk 2]
```diff
@@ -839,8 +841,18 @@ public float MaxArtificialHealth
...
+            set => humeshieldStat.CurValue = value;
+        }
+
+        /// <summary>
+        /// Gets or sets the player's Hume Shield.
+        /// </summary>
+        /// <remarks>This value can bypass the role's hume shield maximum. However, this value will only be visible to the end-player as Hume Shield if <see cref="FpcRole.IsHumeShieldedRole"/> is <see langword="true"/>. Otherwise, the game will treat the player as though they have the amount of Hume Shield specified, even though they cannot see it.</remarks>
+        public float MaxHumeShield
+        {
+            get => humeshieldStat.CurValue;
```
Comment (Thundermaker300) [inline Exiled.API/Features/Player.cs:854]:
```suggestion
            get => humeshieldStat.MaxValue;
```

File: Exiled.API/Features/Player.cs [hunk 3]
```diff
@@ -839,8 +841,18 @@ public float MaxArtificialHealth
...
+            set => humeshieldStat.CurValue = value;
```
Comment (Thundermaker300) [inline Exiled.API/Features/Player.cs:855]:
```suggestion
            set => humeshieldStat.CustomMaxValue = value;
```

**Review by Thundermaker300 — CHANGES_REQUESTED** [review]

**Review by Thundermaker300 — APPROVED** [review]

File: Exiled.API/Features/Player.cs [hunk 4]
```diff
@@ -105,6 +105,7 @@ public class Player : IEntity
 
         private ReferenceHub referenceHub;
         private CustomHealthStat healthStat;
+        private CustomHumeShieldStat humeshieldStat;
```
Comment (Thundermaker300) [inline Exiled.API/Features/Player.cs:108]: Should probably be named `humeShieldStat`, to follow base-game naming scheme of `HumeShield[...]`

Comment (louis1706) [reply Exiled.API/Features/Player.cs:113]: i have followed what Exiled use for name also it's private so i can't start with H but with h only

Comment (Thundermaker300) [reply Exiled.API/Features/Player.cs:113]: Sorry, I meant the capital `S` in shield, not the capital H (edited to reflect). The private variable should be named `humeShieldStat`, not `humeshieldStat`.

File: Exiled.Events/Patches/Generic/PlayerMaxHumeShield.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    internal static class PlayerMaxHumeShield
+    {
+        private static bool Postfix(DynamicHumeShieldController __instance, ref float __result)
+        {
+            Player player = Player.Get(__instance.Owner);
+            if (player.MaxHS == default)
+                __result = player.MaxHS;
+            else
+                __result = __instance.ShieldOverHealth.Evaluate(__instance._hp.NormalizedValue);
+            return false;
```
Comment (iopietro) [inline Exiled.Events/Patches/Generic/PlayerMaxHumeShield.cs:29]:
```suggestion
            Player player = Player.Get(__instance.Owner);

            if (player.MaxHS == default)
                __result = player.MaxHS;
            else
                __result = __instance.ShieldOverHealth.Evaluate(__instance._hp.NormalizedValue);

            return false;
```

File: Exiled.API/Features/CustomHealthStat.cs [hunk 6]
```diff
@@ -21,5 +21,13 @@ public class CustomHealthStat : HealthStat
         /// Gets or sets the maximum amount of health the player will have.
         /// </summary>
         public float CustomMaxValue { get; set; }
+
+        /// <inheritdoc/>
+        public override void ClassChanged()
+        {
+            base.ClassChanged();
+
+            CustomMaxValue = 0f;
```
Comment (iopietro) [inline Exiled.API/Features/CustomHealthStat.cs:30]: Should we remove `player::MaxHealth = default` in `ChangingRole::UpdatePlayerRole` method?

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.API/Features/CustomHealthStat.cs (hunk 6 above)
Comment (louis1706) [reply Exiled.API/Features/CustomHealthStat.cs:30]: Yeah

**Pull request closed without merging**